### PR TITLE
refactored subclasses

### PR DESCRIPTION
**Pull request by VolodymyrWork23**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -46,27 +46,28 @@ export function fromWrappedInstanceMethod<
   return function(instance: TInstance) {
     // tslint:disable-next-line:only-arrow-functions
     return Promise.resolve(function(arg: TIn): Promise<TOut> {
```
Comment (captaincaius) [inline src/index.ts]: what is this?  I don't understand it

File: src/mongoose.ts [hunk 2]
```diff
@@ -111,89 +111,60 @@ export function htMongooseFactory(mongoose: any) {
...
+  function WithUpdateByBody(entityToUpdate: string, updateWithEntity: string) {
+    return WithDoWork(async requestData => {
+      if (Object.keys(requestData).includes(entityToUpdate)) {
+        await requestData[entityToUpdate].set(requestData[updateWithEntity]);
+        return await requestData[entityToUpdate].save();
+      } else {
+        throw Boom.badRequest('Resource not found');
+      }
+    });
   }
```
Comment (captaincaius) [inline src/mongoose.ts]: as mentioned in STAF-2255, the above shouldn't save() - we already have a save.  The customer should just call save separately.

Also please change the names
1. function name to WithUpdateDocument
2. updateWithEntity -> propertyKeyWithNewData
3. entityToUpdate->propertyKeyOfDocument

File: src/subclassers.ts [hunk 3]
```diff
@@ -198,3 +198,29 @@ export function WithAttached<
...
+
+export function WithDoWork<TProjector extends AnyAsyncProjector>(
+  projector: TProjector
+) {
+  // tslint:disable-next-line:only-arrow-functions
+  return function<TSuper extends Constructor>(Super: TSuper) {
+    return class DoWork extends Super {
+      constructor(...args: any[]) {
+        super(...args);
+      }
```
Comment (captaincaius) [inline src/subclassers.ts]: 1. [ ] minor: change class name to WithDoWork
2. [ ] I think all this stuff below is unnecessary b/c executeHipThrustable will do exactly this anyway.  I think you can just await projector(this); Even without the try/catch

File: src/subclassers.ts [hunk 4]
```diff
@@ -198,3 +203,41 @@ export function WithAttached<
     };
   };
 }
```
[the remark this answers is not in the document]
Comment (captaincaius) [reply src/subclassers.ts]: DONE

File: src/mongoose.ts [hunk 5]
```diff
@@ -112,88 +117,46 @@ export function htMongooseFactory(mongoose: any) {
...
-      };
-    };
+    return WithDoWork(entityToUpdate, DoWorkAllowedMethods.UPDATE);
   }
 
   // @note: sanitize body validates modified only!  This is cause you usually will only send fields to update.
   function WithBodySanitized<
     TSafeBody extends ReturnType<TInstance['toObject']>,
     TDocFactory extends DocumentFactory<any>,
     TInstance extends ReturnType<TDocFactory>
```
[the remark this answers is not in the document]
Comment (VolodymyrWork23) [reply src/mongoose.ts]: i decide to remove ? and when i calling this function i pass the DocFactory what returns empty object

File: src/index.ts [hunk 6]
```diff
@@ -46,27 +46,28 @@ export function fromWrappedInstanceMethod<
...
-  };
+  return HTPipe(
+    WithAttached(
+      'defaultNoopAuthUserKey',
+      fromWrappedInstanceMethod('finalAuthForAll'),
+      'isUserExist'
+    ),
+    WithFinalAuth('defaultNoopAuthUserKey', 'isUserExist')
+  );
 }
```
[the remark this answers is not in the document]
Comment (VolodymyrWork23) [reply src/index.ts]: i hope i did it correctly
i wasn't sure that you want to left the old functionality of WithFinalAuth so i did it like others with projector

File: src/index.ts (hunk 1 above)
Comment (VolodymyrWork23) [reply src/index.ts]: Obsolete changes
already removed

File: src/mongoose.ts (hunk 2 above)
Comment (VolodymyrWork23) [reply src/mongoose.ts]: done

File: src/subclassers.ts (hunk 3 above)
Comment (VolodymyrWork23) [reply src/subclassers.ts]: done

File: src/subclassers.ts [hunk 7]
```diff
@@ -121,16 +110,13 @@ export function WithResponse<
...
-export function WithFinalAuth<
-  TPrincipalKey extends string,
-  TAuthKey extends string
->(principalKey: TPrincipalKey, authorizerKey: TAuthKey) {
+export function WithFinalAuth<TProjector extends AnyAsyncProjector>(
+  projector: TProjector
+) {
   // tslint:disable-next-line:only-arrow-functions
   return function<
     TSuper extends Constructor<
```
Comment (captaincaius) [inline src/subclassers.ts]: this type isn't right - in this case, TSuper doesn't have to be constrained cause we don't need anything from it.
Look up what Record<> does.
[ ] You probably just want to return function<TSuper> and not put any constraints on it.

Comment (captaincaius) [reply src/subclassers.ts]: actually ... you DO want the superclass's instance ("this") to be compatible with the projector.  So here's try 2:
`return function<TSuper extends Parameter<TProjector>[0]>`
something like that
- [ ] Constrain the superclass to be compatible with the projector's parameter.

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (captaincaius) [reply src/index.ts]: yup on the javascript side this is right, but the typescript side needs love.  I put some tasks below too, but here we're missing some constraints, which unfortunately takes us back to maybe needing to copypasta like WithAttached :(
- [ ] pending

File: src/mongoose.ts [hunk 8]
```diff
@@ -111,48 +112,35 @@ export function htMongooseFactory(mongoose: any) {
     };
   }
```
Comment (captaincaius) [inline src/mongoose.ts]: i just realized all these Object.keys(...).includes(...) everywhere - it's pretty costly considering it's only there to protect a developer from shooting themselves.
I'd say change them all to
- [ ] `if(requestData[propertyKeyOfDocument]) {` everywhere

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (captaincaius) [reply src/index.ts]: actually i just realized that typescript MIGHT already be smart enough to complain if you try to give it a class that doesn't have something that should work... try the following test cases...
`
class SomeSuper {
  user = {name: "Link"};
  goodAuthorizer = (testPrincipal: {name: string}) => testPrincipal.name=="Link";
  goddessAuthorizer = (testPrincipal: {name: string}) => testPrincipal.name=="Zelda";
  needyAuthorizer = (testPrincipal: {name: string, power: number}) => true;
  notEvenAuthorizer = true;
}
`
1. `WithFinalAuthFrom('user', 'goodAuthorizer')(SomeSuper);` // should work
2. `WithFinalAuthFrom('user', 'goddessAuthorizer')(SomeClass);` // should also work
3. `WithFinalAuthFrom('user', 'nonexistentAuthorizer')(SomeClass);` // should not compile
4. `WithFinalAuthFrom('user', 'notEvenAuthorizer')(SomeClass);` // should not compile
5. `WithFinalAuthFrom('user', 'needyAuthorizer')(SomeClass);` // should not compile
6. `WithFinalAuthFrom('notUser', 'goodAuthorizer')(SomeSuper);` // should not compile

- [ ] Check em and fix if needed

File: src/index.ts [hunk 9]
```diff
@@ -52,18 +49,62 @@ export function fromWrappedInstanceMethod<
...
 
 export function WithNoopPreAuth() {
```
Comment (captaincaius) [inline src/index.ts]: I thought this fake field is no longer necessary...

Comment (captaincaius) [reply src/index.ts]: - [ ] defaultNoopBlah

**Pull request merged**